src/Compilers/CSharp/Portable/Binder/UsingStatementBinder.cs

```diff
@@ -64,66 +64,38 @@ internal override BoundStatement BindUsingStatementParts(DiagnosticBag diagnosti
             bool hasAwait = _syntax.AwaitKeyword.Kind() != default;
 
             Debug.Assert((expressionSyntax == null) ^ (declarationSyntax == null)); // Can't have both or neither.
+            TypeSymbol disposableInterface = getDisposableInterface(hasAwait);
 
-            TypeSymbol iDisposable = hasAwait
-                ? this.Compilation.GetWellKnownType(WellKnownType.System_IAsyncDisposable)
-                : this.Compilation.GetSpecialType(SpecialType.System_IDisposable);
-
-            Debug.Assert((object)iDisposable != null);
-            bool hasErrors = ReportUseSiteDiagnostics(iDisposable, diagnostics, hasAwait ? _syntax.AwaitKeyword : _syntax.UsingKeyword);
+            Debug.Assert((object)disposableInterface != null);
+            bool hasErrors = ReportUseSiteDiagnostics(disposableInterface, diagnostics, hasAwait ? _syntax.AwaitKeyword : _syntax.UsingKeyword);
 
             Conversion iDisposableConversion = Conversion.NoConversion;
             BoundMultipleLocalDeclarations declarationsOpt = null;
             BoundExpression expressionOpt = null;
             AwaitableInfo awaitOpt = null;
+            TypeSymbol declarationTypeOpt = null;
+
             if (expressionSyntax != null)
             {
                 expressionOpt = this.BindTargetExpression(diagnostics, originalBinder);
-
-                HashSet<DiagnosticInfo> useSiteDiagnostics = null;
-                iDisposableConversion = originalBinder.Conversions.ClassifyImplicitConversionFromExpression(expressionOpt, iDisposable, ref useSiteDiagnostics);
-                diagnostics.Add(expressionSyntax, useSiteDiagnostics);
-
-                if (!iDisposableConversion.IsImplicit)
-                {
-                    TypeSymbol expressionType = expressionOpt.Type;
-                    if ((object)expressionType == null || !expressionType.IsErrorType())
-                    {
-                        Error(diagnostics, hasAwait ? ErrorCode.ERR_NoConvToIAsyncDisp : ErrorCode.ERR_NoConvToIDisp, expressionSyntax, expressionOpt.Display);
-                    }
-                    hasErrors = true;
-                }
+                hasErrors |= !initConversion(fromExpression: true);
             }
             else
             {
                 ImmutableArray<BoundLocalDeclaration> declarations;
                 originalBinder.BindForOrUsingOrFixedDeclarations(declarationSyntax, LocalDeclarationKind.UsingVariable, diagnostics, out declarations);
 
                 Debug.Assert(!declarations.IsEmpty);
-
                 declarationsOpt = new BoundMultipleLocalDeclarations(declarationSyntax, declarations);
+                declarationTypeOpt = declarations[0].DeclaredType.Type;
 
-                TypeSymbol declType = declarations[0].DeclaredType.Type;
-
-                if (declType.IsDynamic())
+                if (declarationTypeOpt.IsDynamic())
```

Comment:
>declarationTypeOpt [](start = 20, length = 18)

If this is optional, consider `?.` instead. If it is not optional, consider renaming.

Reply:
It's optional in the method, but it's always set in this branch, so I don't see a great way. I could have an extra local for the branch, but that seems overkill :-(

---
In reply to: [232006228](https://github.com/dotnet/roslyn/pull/30363#discussion_r232006228) [](ancestors = 232006228)

Reply:
You could assert, but I'm ok with this.

---
In reply to: [232067572](https://github.com/dotnet/roslyn/pull/30363#discussion_r232067572) [](ancestors = 232067572,232006228)